orttraining/orttraining/test/python/orttraining_test_backend_api.py

```diff
@@ -0,0 +1,161 @@
+    
+    # generate model and optimizer value from a training instance
+    init_model_state, init_opt_state = generate_model_optimizer_from_training_instance(device, opts_dict)
+    opts = orttrainer.ORTTrainerOptions(opts_dict)
+    optim_config = optim.LambConfig(lr=learning_rate)
+    model, model_desc, loss_fn, batcher_fn, train_data, _, _ = _load_pytorch_transformer_model(device)
+    trainer = orttrainer.ORTTrainer(model, model_desc, optim_config, loss_fn=loss_fn, options=opts)
+
+    #train(trainer, train_data, batcher_fn)
+    train(trainer, train_data, batcher_fn)
+
+    trainer._training_session.load_model_optimizer_state(init_model_state, init_opt_state, False)
+
+    # train one step
+    train(trainer, train_data, batcher_fn, 1)
+    
+    verify_model_state(trainer, init_model_state)
+    
+    verify_opt_state(trainer, init_opt_state)
+
+    verify_part_info(trainer, False)
+
+@distributed_setup
+def test_zero(world_rank, world_size, device, checkpoint_dir = 'checkpoint_dir/distributed_zero/mixed_precision/lamb/'):
+    #print("in test_zero")
+    learning_rate = 1e-10
+    seed = 1
+    torch.manual_seed(seed)
+    set_seed(seed)
+    opts_dict = {
+                'device' : {'id' : device},
+                'mixed_precision':
+                {
+                    'enabled': True
+                },
+                'distributed' :
+                {
+                    'world_rank' : world_rank,
+                    'world_size' : world_size,
+                    'allreduce_post_accumulation' : True,
+                    'deepspeed_zero_optimization':
+                    {
+                        'stage': 1
+                    }
+                },
+                'debug' : {'deterministic_compute': True}
+            }
+    
+    # generate model and optimizer value from a training instance
+    init_model_state, init_opt_state = generate_model_optimizer_from_training_instance(device, opts_dict)
+    opts = orttrainer.ORTTrainerOptions(opts_dict)
+    optim_config = optim.LambConfig(lr=learning_rate)
+    model, model_desc, loss_fn, batcher_fn, train_data, _, _ = _load_pytorch_transformer_model(device)
+    trainer = orttrainer.ORTTrainer(model, model_desc, optim_config, loss_fn=loss_fn, options=opts)
+
+    #train(trainer, train_data, batcher_fn)
+    train(trainer, train_data, batcher_fn)
+
+    trainer._training_session.load_model_optimizer_state(init_model_state, init_opt_state, False)
+
+    # train one step
+    train(trainer, train_data, batcher_fn, 1)
+
+    actual_model_state = trainer._training_session.get_model_state(include_mixed_precision_weights=True)
+    
+    verify_model_state(trainer, init_model_state)
+    
+    verify_opt_state(trainer, init_opt_state)
+
+    part_info = trainer._training_session.get_partition_info_map()
+
+    verify_part_info(trainer, True)
+
+#test_backend_api()
+test_zero(checkpoint_dir = 'checkpoint_dir/distributed_zero/mixed_precision/lamb/')
```

Comment:
we need to test atleast two scenarios, one simple non-model parallel scenario where we run a train step, get model and optimizer states and ensure that they are present for all original weights(found in the input onnx model). The second one for Zero, where we must get the model and optimizer states, make sure that they are correct. this can be validated in different ways, but one of them is to save them on different workers, then load them on a single one, and make sure all the info returned is correct. i.e. all the partitioned states when aggregated form a complete state corresponding to the original onnx/torch model.

Reply:
I have local tests for these two scenarios. Will expand test coverage in next PR